java/Ice/maxConnections failure (debug Ubuntu)

From
https://github.com/zeroc-ice/ice/actions/runs/11261761158/job/31315911872?pr=2879

```
*** [229/305] Running java/Ice/maxConnections tests ***
[ running client/server test - 10/09/24 19:59:27 ]
- Config: ssl
(/opt/hostedtoolcache/Java_Oracle_jdk/17/x64/bin/java -ea -Djdk.net.URLClassPath.disableClassPathURLCheck=true -Dtest.class=test.Ice.maxConnections.Server test.TestDriver --Ice.Default.Host=127.0.0.1 --Test.BasePort=14200 --Ice.Warn.Connections=1 --Ice.Default.Protocol=ssl --Ice.IPv6=0 --IceSSL.*** --IceSSL.DefaultDir=/home/runner/work/ice/ice/certs --IceSSL.Keystore=server.jks --Ice.ThreadPool.Server.Size=1 --Ice.ThreadPool.Server.SizeMax=3 --Ice.ThreadPool.Server.SizeWarn=0 --Ice.PrintAdapterReady=1 --Ice.Trace.Network=2 --Ice.LogFile=/home/runner/work/ice/ice/java/test/src/main/java/test/Ice/maxConnections/server-100924-1959.log env={'CLASSPATH': '/home/runner/work/ice/ice/java/lib/test.jar'})
(/opt/hostedtoolcache/Java_Oracle_jdk/17/x64/bin/java -ea -Djdk.net.URLClassPath.disableClassPathURLCheck=true -Dtest.class=test.Ice.maxConnections.Client test.TestDriver --Ice.Default.Host=127.0.0.1 --Test.BasePort=14200 --Ice.Warn.Connections=1 --Ice.Default.Protocol=ssl --Ice.IPv6=0 --IceSSL.*** --IceSSL.DefaultDir=/home/runner/work/ice/ice/certs --IceSSL.Keystore=client.jks --Ice.Trace.Network=2 --Ice.LogFile=/home/runner/work/ice/ice/java/test/src/main/java/test/Ice/maxConnections/client-100924-1959.log env={'CLASSPATH': '/home/runner/work/ice/ice/java/lib/test.jar'})
testing the creation of 100 connections... ok
testing the creation of 10 connections with connection lost at 11... ok
testing the creation of 10 connections with connection lost at 11 then recovery... ok
[ running client/server test - 10/09/24 19:59:31 ]
- Config: ws,ipv6,serialize,mx
(/opt/hostedtoolcache/Java_Oracle_jdk/17/x64/bin/java -ea -Djdk.net.URLClassPath.disableClassPathURLCheck=true -Dtest.class=test.Ice.maxConnections.Server test.TestDriver --Ice.Default.Host=::1 --Test.BasePort=14200 --Ice.Warn.Connections=1 --Ice.Default.Protocol=ws --Ice.ThreadPool.Server.Serialize=1 --Ice.IPv6=1 --Ice.PreferIPv6Address=1 --Ice.Admin.Endpoints="tcp -h \"::1\"" --Ice.Admin.InstanceName=Server --IceMX.Metrics.Debug.GroupBy=id --IceMX.Metrics.Parent.GroupBy=parent --IceMX.Metrics.All.GroupBy=none --Ice.ThreadPool.Server.Size=1 --Ice.ThreadPool.Server.SizeMax=3 --Ice.ThreadPool.Server.SizeWarn=0 --Ice.PrintAdapterReady=1 --Ice.Trace.Network=2 --Ice.LogFile=/home/runner/work/ice/ice/java/test/src/main/java/test/Ice/maxConnections/server-100924-1959.log env={'CLASSPATH': '/home/runner/work/ice/ice/java/lib/test.jar'})
(/opt/hostedtoolcache/Java_Oracle_jdk/17/x64/bin/java -ea -Djdk.net.URLClassPath.disableClassPathURLCheck=true -Dtest.class=test.Ice.maxConnections.Client test.TestDriver --Ice.Default.Host=::1 --Test.BasePort=14200 --Ice.Warn.Connections=1 --Ice.Default.Protocol=ws --Ice.ThreadPool.Server.Serialize=1 --Ice.IPv6=1 --Ice.PreferIPv6Address=1 --Ice.Admin.Endpoints="tcp -h \"::1\"" --Ice.Admin.InstanceName=Client --IceMX.Metrics.Debug.GroupBy=id --IceMX.Metrics.Parent.GroupBy=parent --IceMX.Metrics.All.GroupBy=none --Ice.Trace.Network=2 --Ice.LogFile=/home/runner/work/ice/ice/java/test/src/main/java/test/Ice/maxConnections/client-100924-1959.log env={'CLASSPATH': '/home/runner/work/ice/ice/java/lib/test.jar'})
testing the creation of 100 connections... ok
testing the creation of 10 connections with connection lost at 11... ok
testing the creation of 10 connections with connection lost at 11 then recovery... com.zeroc.Ice.ConnectionLostException
	at com.zeroc.Ice.OutgoingAsync.waitForResponseOrUserEx(OutgoingAsync.java:110)
	at com.zeroc.Ice.OutgoingAsync.waitForResponse(OutgoingAsync.java:93)
	at com.zeroc.Ice._ObjectPrxI.ice_ping(_ObjectPrxI.java:59)
	at com.zeroc.Ice._ObjectPrxFactoryMethods.ice_ping(_ObjectPrxFactoryMethods.java:11)
	at com.zeroc.Ice._ObjectPrxI.ice_ping(_ObjectPrxI.java:55)
	at com.zeroc.Ice._ObjectPrxFactoryMethods.ice_ping(_ObjectPrxFactoryMethods.java:11)
	at test.Ice.maxConnections.AllTests.testCreateConnectionsWithMaxAndRecovery(AllTests.java:111)
	at test.Ice.maxConnections.AllTests.allTests(AllTests.java:21)
	at test.Ice.maxConnections.Client.run(Client.java:12)
	at test.TestDriver.main(TestDriver.java:26)
Caused by: java.net.SocketException: Connection reset
	at java.base/sun.nio.ch.SocketChannelImpl.throwConnectionReset(SocketChannelImpl.java:394)
	at java.base/sun.nio.ch.SocketChannelImpl.read(SocketChannelImpl.java:426)
	at com.zeroc.Ice.StreamSocket.read(StreamSocket.java:144)
	at com.zeroc.Ice.StreamSocket.read(StreamSocket.java:116)
	at com.zeroc.Ice.TcpTransceiver.read(TcpTransceiver.java:49)
	at com.zeroc.Ice.WSTransceiver.read(WSTransceiver.java:333)
	at com.zeroc.Ice.IdleTimeoutTransceiverDecorator.read(IdleTimeoutTransceiverDecorator.java:82)
	at com.zeroc.Ice.ConnectionI.read(ConnectionI.java:2450)
	at com.zeroc.Ice.ConnectionI.message(ConnectionI.java:564)
	at com.zeroc.Ice.ThreadPool.run(ThreadPool.java:352)
	at com.zeroc.Ice.ThreadPool$EventHandlerThread.run(ThreadPool.java:662)
	at java.base/java.lang.Thread.run(Thread.java:842)
saved /home/runner/work/ice/ice/java/test/src/main/java/test/Ice/maxConnections/client-100924-1959.log
saved /home/runner/work/ice/ice/java/test/src/main/java/test/Ice/maxConnections/server-100924-1959.log

```

--- a/cpp/test/Ice/maxConnections/AllTests.cpp
+++ b/cpp/test/Ice/maxConnections/AllTests.cpp
@@ -13,7 +13,7 @@ using namespace Test;
 
 // Verify that we can create connectionCount connections and send a ping on each connection.
 void
-testCreateConnections(TestIntfPrx p, int connectionCount)
+testCreateConnections(TestIntfPrx p, int connectionCount, function<void()> postCloseDelay)
 {
     cout << "testing the creation of " << connectionCount << " connections... " << flush;
     vector<ConnectionPtr> connectionList;
@@ -29,12 +29,18 @@ testCreateConnections(TestIntfPrx p, int connectionCount)
     {
         connection->close().get();
     }
+
+    if (postCloseDelay)
+    {
+        postCloseDelay();
+    }
+
     cout << "ok" << endl;
 }
 
 // Verify that we can create connectionCount connections and send a ping on each connection.
 void
-testCreateConnectionsWithMax(TestIntfPrx p, int max)
+testCreateConnectionsWithMax(TestIntfPrx p, int max, function<void()> postCloseDelay)
 {
     cout << "testing the creation of " << max << " connections with connection lost at " << (max + 1) << "... "
          << flush;
@@ -63,12 +69,18 @@ testCreateConnectionsWithMax(TestIntfPrx p, int max)
     {
         connection->close().get();
     }
+
+    if (postCloseDelay)
+    {
+        postCloseDelay();
+    }
+
     cout << "ok" << endl;
 }
 
 // Verify that we can create max connections, then connection lost, then recover.
 void
-testCreateConnectionsWithMaxAndRecovery(TestIntfPrx p, int max)
+testCreateConnectionsWithMaxAndRecovery(TestIntfPrx p, int max, function<void()> postCloseDelay)
 {
     cout << "testing the creation of " << max << " connections with connection lost at " << (max + 1)
          << " then recovery... " << flush;
@@ -96,6 +108,11 @@ testCreateConnectionsWithMaxAndRecovery(TestIntfPrx p, int max)
     connectionList.front()->close().get();
     connectionList.erase(connectionList.begin());
 
+    if (postCloseDelay)
+    {
+        postCloseDelay();
+    }
+
     // Try again
     try
     {
@@ -114,6 +131,11 @@ testCreateConnectionsWithMaxAndRecovery(TestIntfPrx p, int max)
     {
         connection->close().get();
     }
+
+    if (postCloseDelay)
+    {
+        postCloseDelay();
+    }
     cout << "ok" << endl;
 }
 
@@ -127,9 +149,19 @@ allTests(TestHelper* helper)
     string proxyStringMax10 = "test: " + helper->getTestEndpoint(1);
     TestIntfPrx pMax10(communicator, proxyStringMax10);
 
-    testCreateConnections(p, 100);
-    testCreateConnectionsWithMax(pMax10, 10);
-    testCreateConnectionsWithMaxAndRecovery(pMax10, 10);
+    string transport = helper->getTestProtocol();
+
+    // When the transport is WS or WSS, we need to wait a little bit: the server closes the connection after it
+    // gets a transport frame from the client.
+    function<void()> postCloseDelay;
+    if (transport.find("ws") == 0)
+    {
+        postCloseDelay = []() { this_thread::sleep_for(50ms); };
+    }
+
+    testCreateConnections(p, 100, postCloseDelay);
+    testCreateConnectionsWithMax(pMax10, 10, postCloseDelay);
+    testCreateConnectionsWithMaxAndRecovery(pMax10, 10, postCloseDelay);
 
     p->shutdown();
 }

--- a/cpp/test/Ice/maxConnections/Client.cpp
+++ b/cpp/test/Ice/maxConnections/Client.cpp
@@ -16,7 +16,11 @@ public:
 void
 Client::run(int argc, char** argv)
 {
-    Ice::CommunicatorHolder communicator = initialize(argc, argv);
+    auto properties = createTestProperties(argc, argv);
+    // We disable retries to make the logs clearer and avoid hiding potential issues.
+    properties->setProperty("Ice.RetryIntervals", "-1");
+
+    Ice::CommunicatorHolder communicator = initialize(argc, argv, properties);
 
     void allTests(Test::TestHelper*);
     allTests(this);

--- a/csharp/test/Ice/maxConnections/AllTests.cs
+++ b/csharp/test/Ice/maxConnections/AllTests.cs
@@ -15,15 +15,27 @@ internal class AllTests : global::Test.AllTests
         string proxyStringMax10 = $"test: {helper.getTestEndpoint(1)}";
         Test.TestIntfPrx pMax10 = Test.TestIntfPrxHelper.createProxy(communicator, proxyStringMax10);
 
-        await testCreateConnections(p, 100, helper.getWriter());
-        await testCreateConnectionsWithMax(pMax10, 10, helper.getWriter());
-        await testCreateConnectionsWithMaxAndRecovery(pMax10, 10, helper.getWriter());
+        // When the transport is WS or WSS, we need to wait a little bit: the server closes the connection after it
+        // gets a transport frame from the client.
+        Func<Task>? postCloseDelay = null;
+        if (helper.getTestProtocol().StartsWith("ws"))
+        {
+            postCloseDelay = () => Task.Delay(TimeSpan.FromMilliseconds(50));
+        }
+
+        await testCreateConnections(p, 100, helper.getWriter(), postCloseDelay);
+        await testCreateConnectionsWithMax(pMax10, 10, helper.getWriter(), postCloseDelay);
+        await testCreateConnectionsWithMaxAndRecovery(pMax10, 10, helper.getWriter(), postCloseDelay);
 
         await p.shutdownAsync();
     }
 
     // Verify that we can create connectionCount connections and send a ping on each connection.
-    private static async Task testCreateConnections(Test.TestIntfPrx p, int connectionCount, TextWriter output)
+    private static async Task testCreateConnections(
+        Test.TestIntfPrx p,
+        int connectionCount,
+        TextWriter output,
+        Func<Task>? postCloseDelay)
     {
         output.Write($"testing the creation of {connectionCount} connections... ");
         output.Flush();
@@ -38,11 +50,20 @@ internal class AllTests : global::Test.AllTests
 
         // Close all connections
         await Task.WhenAll(connectionList.Select(c => c.closeAsync()));
+
+        if (postCloseDelay is not null)
+        {
+            await postCloseDelay();
+        }
         output.WriteLine("ok");
     }
 
     // Verify that we can create max connections but not more.
-    private static async Task testCreateConnectionsWithMax(Test.TestIntfPrx p, int max, TextWriter output)
+    private static async Task testCreateConnectionsWithMax(
+        Test.TestIntfPrx p,
+        int max,
+        TextWriter output,
+        Func<Task>? postCloseDelay)
     {
         output.Write($"testing the creation of {max} connections with connection lost at {max + 1}... ");
         output.Flush();
@@ -68,11 +89,20 @@ internal class AllTests : global::Test.AllTests
 
         // Close all connections
         await Task.WhenAll(connectionList.Select(c => c.closeAsync()));
+
+        if (postCloseDelay is not null)
+        {
+            await postCloseDelay();
+        }
         output.WriteLine("ok");
     }
 
     // Verify that we can create max connections, then connection lost, then recover.
-    private static async Task testCreateConnectionsWithMaxAndRecovery(Test.TestIntfPrx p, int max, TextWriter output)
+    private static async Task testCreateConnectionsWithMaxAndRecovery(
+        Test.TestIntfPrx p,
+        int max,
+        TextWriter output,
+        Func<Task>? postCloseDelay)
     {
         output.Write($"testing the creation of {max} connections with connection lost at {max + 1} then recovery... ");
         output.Flush();
@@ -100,12 +130,22 @@ internal class AllTests : global::Test.AllTests
         await connectionList[0].closeAsync();
         connectionList.RemoveAt(0);
 
+        if (postCloseDelay is not null)
+        {
+            await postCloseDelay();
+        }
+
         // Try again
         await p.ice_pingAsync();
         connectionList.Add(p.ice_getCachedConnection()!);
 
         // Close all connections
         await Task.WhenAll(connectionList.Select(c => c.closeAsync()));
+
+        if (postCloseDelay is not null)
+        {
+            await postCloseDelay();
+        }
         output.WriteLine("ok");
     }
 }

--- a/csharp/test/Ice/maxConnections/Client.cs
+++ b/csharp/test/Ice/maxConnections/Client.cs
@@ -8,7 +8,11 @@ public class Client : global::Test.TestHelper
 {
     public override async Task runAsync(string[] args)
     {
-        using var communicator = initialize(ref args);
+        Properties properties = createTestProperties(ref args);
+        // We disable retries to make the logs clearer and avoid hiding potential issues.
+        properties.setProperty("Ice.RetryIntervals", "-1");
+
+        using var communicator = initialize(properties);
         await AllTests.allTests(this);
     }
 

--- a/java/test/src/main/java/test/Ice/maxConnections/AllTests.java
+++ b/java/test/src/main/java/test/Ice/maxConnections/AllTests.java
@@ -16,16 +16,27 @@ public class AllTests {
         String proxyStringMax10 = "test: " + helper.getTestEndpoint(1);
         var pMax10 = TestIntfPrx.createProxy(communicator, proxyStringMax10);
 
-        testCreateConnections(p, 100, helper.getWriter());
-        testCreateConnectionsWithMax(pMax10, 10, helper.getWriter());
-        testCreateConnectionsWithMaxAndRecovery(pMax10, 10, helper.getWriter());
+        Runnable postCloseDelay = null;
+        // When the transport is WS or WSS, we need to wait a little bit: the server closes
+        // the connection after it gets a transport frame from the client.
+        if (helper.getTestProtocol().startsWith("ws")) {
+            try {
+                Thread.sleep(50);
+            } catch (InterruptedException ex) {
+                // ignore
+            }
+        }
+
+        testCreateConnections(p, 100, helper.getWriter(), postCloseDelay);
+        testCreateConnectionsWithMax(pMax10, 10, helper.getWriter(), postCloseDelay);
+        testCreateConnectionsWithMaxAndRecovery(pMax10, 10, helper.getWriter(), postCloseDelay);
 
         p.shutdown();
     }
 
     // Verify that we can create connectionCount connections and send a ping on each connection.
     private static void testCreateConnections(
-            TestIntfPrx p, int connectionCount, PrintWriter output) {
+            TestIntfPrx p, int connectionCount, PrintWriter output, Runnable postCloseDelay) {
         output.write("testing the creation of " + connectionCount + " connections... ");
         output.flush();
 
@@ -41,11 +52,16 @@ public class AllTests {
             connectionList.get(i).close();
         }
 
+        if (postCloseDelay != null) {
+            postCloseDelay.run();
+        }
+
         output.println("ok");
     }
 
     // Verify that we can create max connections but not more.
-    private static void testCreateConnectionsWithMax(TestIntfPrx p, int max, PrintWriter output) {
+    private static void testCreateConnectionsWithMax(
+            TestIntfPrx p, int max, PrintWriter output, Runnable postCloseDelay) {
         output.write(
                 "testing the creation of "
                         + max
@@ -74,12 +90,16 @@ public class AllTests {
             connectionList.get(i).close();
         }
 
+        if (postCloseDelay != null) {
+            postCloseDelay.run();
+        }
+
         output.println("ok");
     }
 
     // Verify that we can create max connections, then connection lost, then recover.
     private static void testCreateConnectionsWithMaxAndRecovery(
-            TestIntfPrx p, int max, PrintWriter output) {
+            TestIntfPrx p, int max, PrintWriter output, Runnable postCloseDelay) {
         output.write(
                 "testing the creation of "
                         + max
@@ -107,6 +127,10 @@ public class AllTests {
         connectionList.get(0).close();
         connectionList.remove(0);
 
+        if (postCloseDelay != null) {
+            postCloseDelay.run();
+        }
+
         // Try again
         p.ice_ping();
         connectionList.add(p.ice_getCachedConnection());
@@ -114,6 +138,10 @@ public class AllTests {
         // Close all connections
         for (int i = 0; i < max; i++) {
             connectionList.get(i).close();
+        }
+
+        if (postCloseDelay != null) {
+            postCloseDelay.run();
         }
 
         output.println("ok");

--- a/java/test/src/main/java/test/Ice/maxConnections/Client.java
+++ b/java/test/src/main/java/test/Ice/maxConnections/Client.java
@@ -8,6 +8,9 @@ public class Client extends test.TestHelper {
         var properties = createTestProperties(args);
         properties.setProperty("Ice.Package.Test", "test.Ice.maxConnections");
 
+        // We disable retries to make the logs clearer and avoid hiding potential issues.
+        properties.setProperty("Ice.RetryIntervals", "-1");
+
         try (var communicator = initialize(properties)) {
             AllTests.allTests(this);
         }

